### PR TITLE
Adds TensorCollection

### LINKED ISSUE #435: Parameter Count
When training large models, it's quite useful to know the size of the model being trained. In pytorch it's fairly straightforward to calculate this:
```python
params = sum(p.numel() for p in model.parameters() if (p.requires_grad or not trainable_only))
```
This is doable because we can use the `model.parameters()` to get a list of all the `Parameter`s in the model, and then sum over the trainable ones.

Currently in dfdx I made a trait for counting parameters called `CountParams`:
```rust
pub trait CountParams {
    fn param_count(&self) -> usize;
}
```
which gets implemented on all tensors like so:
```rust
impl <
    S: Shape,
    E: Dtype + Unit,
    T,
    D: DeviceStorage + ZerosTensor<E> + CopySlice<E>
> CountParams for Tensor<S, E, D, T> {
    fn param_count(&self) -> usize {
        self.shape().num_elements()
    }
}
```

Of course the straightforward way to use this is to implement it on all modules, and then just call `.param_count()` on your model. I don't really think this is a great way to do it though. It seems for every new piece of functionality, we need to constantly implement new traits for each and every module, and as the number of modules grows, this becomes harder and harder to maintain.

One alternative might be to have a state-dict style structure like pytorch has. Something like:
```rust
trait ToStateDict {
    fn state_dict(&self) -> HashMap<String, &Tensor>;
}
```
which would allow modules to specify their storage layout. Once implemented, this could give us CountParams, as well as SaveToNpz for free, because we now have the data and the paths for each param.If the state_dict returned `HashMap<String, &mut Tensor>` instead, it would also give us LoadFromNpz for free as well, because we have the paths as well as the tensors to load into.

Not sure this is the best way, but I don't think constantly implementing new traits will be sustainable long-term.

### PR DESCRIPTION
**Pull request by coreylowman**
Resolves #435 
Related to #460 

- Removes trait GradientUpdate
- Removes trait ParamUpdater
- optimizers now implement VisitTensors and use RecursiveWalker in their Optimizer impl
- Moves trait ResetParams to use TensorCollection, and removes all nn impls
- Adds NumParams using TensorCollection
- Refactor SaveToNpz/LoadFromNpz to use TensorCollection
- Removes all missing gradients tests in nn
    - Now that all nn layers use TensorCollection, testing npz/update/reset_params is enough to cover module walking
    - Unused tensor tests are now added to optimizers to ensure they properly capture unused
- Adds traits VisitsTensors
- Adds trait ModuleVisitor
- Adds trait TensorCollection

TODOs
- [ ] add documentation
- [x] improve use statements
- [x] Move visitors somewhere else other than tensor? top level?

### CODE REVIEW

**Comment by coreylowman** [comment]
@nkoppel thoughts? was able to impl optimizers with this approach as well

File: src/tensor/visitors/base.rs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    {
+        self.path.push(opts.name.into());
+        self.f.visit(self.path.join("."), opts, get_muts(self.m))?;
+        self.path.pop();
+        Ok(())
+    }
+}
+
+impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorMutRef<E, D>> ModuleWalker<M, E, D>
+    for RecursiveWalker<'a, (&'a mut M, &'a M), F>
```
Comment (coreylowman) [inline src/tensor/visitors/base.rs:203]: This impl is what enables EMA. Basically we'd just need to impl VisitTensorMutRef for some struct Ema and add an extension method

File: src/tensor/mod.rs [hunk 2]
```diff
@@ -111,15 +111,13 @@
...
-
 #[cfg(feature = "cuda")]
 pub(crate) mod cuda;
-
 #[cfg(feature = "numpy")]
 pub(crate) mod numpy;
-
 pub(crate) mod storage_traits;
+mod tensor_impls;
+pub(crate) mod visitors; // TODO pub?
```
Comment (coreylowman) [inline src/tensor/mod.rs:120]: note to self

Comment (nkoppel) [reply src/tensor/mod.rs:120]: To me, the visitors module feels like it belongs in `nn`, because its main utility is in operating on nn modules.

Comment (coreylowman) [reply src/tensor/mod.rs:120]: That's what I did initially, however optim now also uses the visitors. optim depending on nn could makes sense the more I think about it...

File: src/tensor/visitors/reset_params.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+    fn try_reset_params(&mut self) -> Result<(), D::Err> {
+        Self::iter_tensors(&mut RecursiveWalker {
+            m: self,
+            f: &mut Resetter,
+            path: &mut Vec::new(),
+        })
+    }
+}
+impl<E: Dtype, D: DeviceStorage, M: TensorCollection<E, D>> ResetParams<E, D> for M {}
```
Comment (nkoppel) [inline src/tensor/visitors/reset_params.rs:31]: This and num_params should be moved to somewhere in `nn` to keep it discoverable in the documentation

File: src/tensor/visitors/base.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    tensor::{DeviceStorage, Tensor},
+};
+
+use std::{string::String, vec::Vec};
+
+pub struct TensorOptions<S: Shape, E: Dtype, D: DeviceStorage> {
+    pub name: &'static str,
+    pub update: bool,
+    pub reset: fn(&mut Tensor<S, E, D>) -> Result<(), D::Err>,
+}
```
[the remark this answers is not in the document]
Comment (coreylowman) [reply src/tensor/visitors/base.rs:14]: 👍 Yeah that's fair and good call out for consistency.

**Comment by nkoppel** [comment]
One thing I don't like about the current implementation is the repeated code in `visitors/base.rs`. A lot of this repetition could be removed by generalizing the behavior of accessing fields in `&T`, `&mut T`, and `(&mut T, &T)` with something like the following:

```rust
pub trait VisitTensors<E: Dtype, D: DeviceStorage> {
    type Argument<'a, M>: VisitTensorArgument<M> where M: 'a;
    type Err;

    fn visit<S: Shape>(
        &mut self,
        full_path: String,
        opts: TensorOptions<S, E, D>,
        t: Self::Argument<'_, Tensor<S, E, D>>,
    ) -> Result<(), Self::Err>;
}

pub trait VisitTensorArgument<M> {
    type WithModule<'a, Mod>: VisitTensorArgument<Mod> where Self: 'a;

    fn get_field<Field, GetRef, GetMut>(
        &mut self,
        get_ref: GetRef,
        get_mut: GetMut,
    ) -> Self::WithModule<'_, Field>
     where
        GetRef: FnMut(&M) -> &Field,
        GetMut: FnMut(&mut M) -> &mut Field,
        Field: TensorCollection<E, D>,
}
```

This would remove the need to have three different `VisitTensor` traits, but would require that we add extra constraints to the `RecursiveWalker` struct to ensure that it has `F::Argument<M>` as it's `m` field. Thoughts?

**Comment by coreylowman** [comment]
I totally agree, and nice sketch. Want to open a PR into this branch with that implemented? I think as long as we can keep that isolated to the implementation for RecursiveWalker (and not infect the user facing TensorCollection/TensorVisitor) it sounds like a great add

File: src/tensor/mod.rs (hunk 2 above)
Comment (coreylowman) [reply src/tensor/mod.rs]: Moved

File: src/nn/mod.rs [hunk 5]
```diff
@@ -104,6 +104,10 @@
 //! mlp.load_state_dict(state_dict)
 //! ```
 
+mod num_params;
+mod reset_params;
+pub mod visitors;
```
Comment (coreylowman) [inline src/nn/mod.rs:109]: Come up with a better name than "visitors". The real purpose for users will be getting access to TensorCollection to impl for modules. Perhaps "collections"?

File: src/nn/visitors.rs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        get_muts: GetMut,
+        name: &str,
+        opts: TensorOptions<S, E, D>,
+    ) -> Result<(), Self::Err>
+    where
+        GetRef: FnMut(&T) -> &Tensor<S, E, D>,
+        GetMut: FnMut(&mut T) -> &mut Tensor<S, E, D>;
+}
+
+pub(crate) struct RecursiveWalker<'a, M, F> {
```
Comment (coreylowman) [inline src/nn/visitors.rs:121]: This should probably be `pub`. Anyone who wants to implement their own optimizer outside of dfdx will need access to this

**Comment by coreylowman** [comment]
@nkoppel I'm pretty happy with where this is, any other organization/naming stuff you think we should do before merge?

File: src/nn/tensor_collection/visitor.rs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    /// Return the view of the tensor
+    fn view<'a, Mod, Field, GetRef, GetMut>(
+        module: &'a mut Self::View<'_, Mod>,
+        get_ref: &mut GetRef,
+        get_mut: &mut GetMut,
+    ) -> Self::View<'a, Field>
+    where
+        GetRef: FnMut(&Mod) -> &Field,
+        GetMut: FnMut(&mut Mod) -> &mut Field;
```
Comment (nkoppel) [inline src/nn/tensor_collection/visitor.rs:47]: I like using 'view' to express the meaning of this trait, but the naming/documentation of the view function isn't very informative.
```suggestion
    /// Given a view to a module, returns a view to one of that module's fields
    fn view_field<'a, Mod, Field, GetRef, GetMut>(
        module: &'a mut Self::View<'_, Mod>,
        get_ref: &mut GetRef,
        get_mut: &mut GetMut,
    ) -> Self::View<'a, Field>
    where
        GetRef: FnMut(&Mod) -> &Field,
        GetMut: FnMut(&mut Mod) -> &mut Field;
```

File: src/nn/tensor_collection/collection.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        )
+    }
+}
+
+/// Options to change behavior of [TensorVisitor]
+pub struct TensorOptions<S: Shape, E: Dtype, D: DeviceStorage> {
+    /// Whether the tensor should be updated with gradients
+    pub do_gradient_update: bool,
+    pub reset: fn(&mut Tensor<S, E, D>) -> Result<(), D::Err>,
+}
```
Comment (nkoppel) [inline src/nn/tensor_collection/collection.rs:63]: I think we should add `#[non_exhaustive]` to this struct so that we can easily add new options in the future if needed. We could also define default values for all fields and use the builder pattern to construct this, but I'm not sure this is better.

**Comment by nkoppel** [comment]
@usbalbin merging this will break parts of #437. You'll needto remove the GradientUpdate implementation and replace it with a TensorCollection implementation.

File: src/nn/tensor_collection/collection.rs (hunk 8 above)
Comment (coreylowman) [reply src/nn/tensor_collection/collection.rs:63]: Oh just kidding, anyone implementing Visitors will need to access those fields. I'll do #[non_exhaustive]!
